### PR TITLE
FIX Jenkins-37962 REGRESSION: parallel karaoke not allowing branch selection or completing correctly

### PR DESCRIPTION
**Pull request by scherler**
# Description

See [JENKINS-37962](https://issues.jenkins-ci.org/browse/JENKINS-37962).

https://youtu.be/JcIC2cP9b6Y

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
https://github.com/jenkinsci/blueocean-acceptance-test/pull/42
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [x] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [x] Run the changes and verified the change matches the issue description
- [x] Reviewed the code
- [x] Verified that the appropriate tests have been written or valid explanation given

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

**Comment by michaelneale** [comment]
@scherler do you know who is running a jenkins configured against this repo? as it is messing up statuses

**Comment by scherler** [comment]
@michaelneale not me, maybe @tfennelly

**Comment by i386** [comment]
Very nice @scherler 👍

**Comment by michaelneale** [comment]
Unfortunately when I try this I can't select either branch now, and it doesn't show anything happening in the parallel section. 

See this



The pipeline to reproduce this: 

```
node {
    stage("hey") {
        sh 'ping -c 5 www.apple.com'

    }
    

    stage("par") {
        parallel (
            "b1" : {
                sh 'ping -c 2 www.apple.com'
            },

            "b2" : {
                sh 'sleep 1 && ping -c 1 www.apple.com'
                sh 'sleep 1 && ping -c 1 www.apple.com'
                sh 'sleep 1 && ping -c 1 www.apple.com'
                sh 'sleep 1 && ping -c 1 www.apple.com'
                sh 'sleep 1 && ping -c 1 www.apple.com'
                sh 'sleep 1 && ping -c 1 www.apple.com'
                sh 'sleep 1 && ping -c 1 www.apple.com'
                sh 'sleep 1 && ping -c 1 www.apple.com'
                sh 'sleep 1 && ping -c 1 www.apple.com'
                sh 'sleep 1 && ping -c 1 www.apple.com'
            } 
            
            
        )
            
    }
    
    
    stage("ho") {
        sh 'ping -c 10 www.apple.com'    
    }
    
}
```

**Comment by michaelneale** [comment]
FYI on non block scoped stages, this works ok. So I think there is something not quite right (perhaps if you can ping @vivek when around, can see what the difference is in the api results).

**Comment by scherler** [comment]
@michaelneale hmm looks like a problem coming from the "flat-line". :( Investigating

File: blueocean-dashboard/src/main/js/redux/actions.js [hunk 1]
```diff
@@ -555,6 +555,7 @@ export const actions = {
                     nodeModel = information.model.filter((item) => item.id === config.node)[0];
                     node = config.node;
                 }
+                // console.log('ACTION_TYPES.SET_NODE', nodeModel);
```
Comment (kzantow) [inline blueocean-dashboard/src/main/js/redux/actions.js]: Maybe use `debug` module?

File: blueocean-dashboard/src/test/json/parallelRunning.json [hunk 2]
```diff
@@ -0,0 +1,329 @@
+[
```
Comment (kzantow) [inline blueocean-dashboard/src/test/json/parallelRunning.json]: Is there no test actually using this file? missed an add?

**Comment by kzantow** [comment]
Is this still a WiP?

**Comment by michaelneale** [comment]
@kzantow yes, some discussion on the ticket, and a new one was spawned: https://issues.jenkins-ci.org/browse/JENKINS-38360 - which has a tentative fix I hope to try later. 

It works fine for non block scoped stages, so hopefully this fills that gap and then it will indeed be complete.

File: blueocean-dashboard/src/test/json/parallelRunning.json (hunk 2 above)
Comment (michaelneale) [inline blueocean-dashboard/src/test/json/parallelRunning.json]: I did wonder what that json is for? cc @scherler

**Comment by michaelneale** [comment]
This is blocked by: https://issues.jenkins-ci.org/browse/JENKINS-38360

@scherler I recommend parking this until that is addressed, and looking at something else.

File: blueocean-dashboard/src/test/json/parallelRunning.json (hunk 2 above)
Comment (scherler) [inline blueocean-dashboard/src/test/json/parallelRunning.json]: I was up to write a junit test but in the end I did not, will remove the file

File: blueocean-dashboard/src/main/js/redux/actions.js (hunk 1 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/redux/actions.js]: Can you point me to an example?

**Comment by michaelneale** [comment]
@scherler as discussed, we can probably move ahead with this as it fixes a known case for classic stages so - tentative 🐝  (and you have ATH for it)

**Comment by michaelneale** [comment]
Very nice!


:bee: from me

File: blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx [hunk 3]
```diff
@@ -286,12 +294,15 @@ export class RunDetailsPipeline extends Component {
                 pathArray.shift();
                 newPath = `${pathArray.join('/')}/`;
             }
+            const isParallel = this.mergedConfig.nodeReducer.edges[0] && nodeInfo.edges[0] ?
+                this.mergedConfig.nodeReducer.edges[0].id === nodeInfo.edges[0].id : false;
```
Comment (tfennelly) [inline blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx]: Would be nice if the above was more human readable.

Comment (tfennelly) [inline blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx]: I read a comment on one other other PRs to the effect that "less code is always better". Naive, to say the least. 

As someone else rightly said (can't remember who atm) ... "Any clown can write code that a computer can understand, but it takes a good programmer to write code that a human can understand". Imo, less code is not always better. The fewest number of easily understood/comprehended lines of code ... sure, that's good.

Comment (tfennelly) [inline blueocean-dashboard/src/main/js/components/RunDetailsPipeline.jsx]: Maybe start by extracting the above code to a function of its own? And thin in there ... break it down into human consumable fragments.

File: blueocean-dashboard/src/main/js/util/logDisplayHelper.js [hunk 4]
```diff
@@ -26,12 +26,28 @@ export const getNodesInformation = (nodes) => {
...
 
+        /*
+         * are we in a node that indicates that we have parallel nodes?
+         */
+        if (item.edges && item.edges.length >= 2) {
+            parallelNodes = item.edges.map((itemParallel) => itemParallel.id);
+        }
+        // in case we had been in a parallel node before, we will indicate it and remove the id of the parallel array
+        const indexParallel = parallelNodes.indexOf(item.id);
+        const isParallel = indexParallel > -1;
```
Comment (tfennelly) [inline blueocean-dashboard/src/main/js/util/logDisplayHelper.js]: NIT: Maybe evaluate in a way that's consistent with the wording of all the docs on `indexOf`. Every doc on this function that I have ever seen says "not equal to -1" (or very similar). From that, I'd always find `const isParallel = indexParallel !== -1;` to be more consistent. `> -1` is obviously not wrong though, so just a NIT 😄

**Comment by tfennelly** [comment]
Sorry ... I selected the wrong option when doing the "review" ... I meant to click "approve" but hit something else.

**Comment by tfennelly** [comment]
@scherler nice and thanks ... 🐝 again in case you'd like it, but I don't think you need it.

File: blueocean-dashboard/src/main/js/util/logDisplayHelper.js (hunk 4 above)
Comment (michaelneale) [inline blueocean-dashboard/src/main/js/util/logDisplayHelper.js]: I am hopig that in future this code can be simplified a lot...

**Pull request merged**